src/Chainweb/Pact/PactService.hs

```diff
@@ -418,40 +422,80 @@ finalizeCheckpointer finalize = do
 _liftCPErr :: Either String a -> PactServiceM cas a
 _liftCPErr = either internalError' return
 
+type Balances  = HM.HashMap Text Decimal
+type GasTotals = HM.HashMap Text Decimal
+
+-- | The principal validation logic for groups of Pact Transactions.
+--
+-- Skips validation for genesis transactions, since gas accounts, etc. don't
+-- exist yet.
+--
 validateChainwebTxs
     :: PactDbEnv'
     -> Checkpointer
     -> BlockCreationTime
     -> BlockHeight
     -> Vector ChainwebTransaction
     -> IO (Vector Bool)
-validateChainwebTxs dbEnv cp blockOriginationTime bh txs =
-    V.mapM checkOne txs
+validateChainwebTxs dbEnv cp blockOriginationTime bh txs
+    | bh == 0 = pure $ V.replicate (V.length txs) False
+    | otherwise = do
+        bs <- balances txs
```

Comment:
If a sender sends in _k_ txs, this will perform _k_ balance checks for that sender, and your `gasTotals` runs two passes over the transactions. I was thinking more something like this:

  - do away with the `GasTotals` map (shouldn't be necessary)
  - implement `checkDebit :: Balances -> Text -> Decimal -> IO (Balances, Bool)` that does a forwarding lookup (i.e. check the map first, then check the db) and fails if the account has insufficient balance -- this ensures one DB lookup per sender. 
  - `validate` becomes `validate :: IORef Balances -> ChainwebTransaction -> IO Bool` -- you could use a fold instead if you want but it's hard to efficiently generate the vector in pure code that way.

Alternatively collect `HashSet Text` for sender first and then do the lookup that way -- the latter has the advantage of being future-proof in the event that we get smart and batch the lookup.

Btw now that my `/send` diagnostic branch is in, the `Bool` probably must be `Mempool.InsertError`

Reply:
> this will perform k balance checks for that sender

Previously it was doing an `IO`-based balance check per TX. This approach here front-loads that effort. Yes it does mean multiple passes over the `Vector` (which is the conversation of a `SubmitBatch`, which we wouldn't expect to be larger than a few thousand TXs at max), but amortized I think the reduction in `IO` calls is an improvement.

I'll think about the other points.